gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java

```diff
@@ -51,29 +73,90 @@ public RecursiveCopyableDataset(final FileSystem fs, Path rootPath, Properties p
     this.pathFilter = DatasetUtils.instantiatePathFilter(properties);
     this.copyableFileFilter = DatasetUtils.instantiateCopyableFileFilter(properties);
     this.glob = glob;
+
+    this.update = Boolean.parseBoolean(properties.getProperty(UPDATE_KEY));
+    this.delete = Boolean.parseBoolean(properties.getProperty(DELETE_KEY));
+    this.deleteEmptyDirectories = Boolean.parseBoolean(properties.getProperty(DELETE_EMPTY_DIRECTORIES_KEY));
+    this.properties = properties;
   }
 
   @Override
   public Collection<? extends CopyEntity> getCopyableFiles(FileSystem targetFs, CopyConfiguration configuration)
       throws IOException {
 
     Path nonGlobSearchPath = PathUtils.deepestNonGlobPath(this.glob);
+    Path targetPath = new Path(configuration.getPublishDir(), PathUtils.relativizePath(this.rootPath, nonGlobSearchPath));
 
-    List<FileStatus> files = FileListUtils.listFilesRecursively(this.fs, this.rootPath, this.pathFilter);
+    Map<Path, FileStatus> filesInSource = createPathMap(getFilesAtPath(this.fs, this.rootPath, this.pathFilter), this.rootPath);
+    Map<Path, FileStatus> filesInTarget = createPathMap(getFilesAtPath(targetFs, targetPath, this.pathFilter), targetPath);
 
-    List<CopyableFile> copyableFiles = Lists.newArrayList();
+    Set<Path> inSourceNotInTarget = Sets.difference(filesInSource.keySet(), filesInTarget.keySet());
```

Comment:
Because we later use those same maps to get the appropriate file statuses for either creating copy work units, determining whether a collision file is the same or not, and scheduling deletes.